[ coco ] 미션3 : 체스판 초기화

choigiseong opened the pull request:
안녕하세요. 백엔드반의 coco라고 합니다.    
부족한 코드를 봐주셔서 감사드립니다.  

점은 class를 나타냈고 숫자는 메소드를 나타냈습니다.

* ##  pieces/Pawn  class
이번 미션에서는 Pawn class 안에 있던 color 상수들을 enum으로 빼서 작성했습니다.   
또 구분문자인 representation을 새로운 필드로 갖게 되었습니다.

* ## pieces/PawnEnum
상수를 대신해 enum을 도입했으며 내부에 구분문자를 가지게 있게 했습니다. 이 때문에  pawn 생성 시 enum 만 파라미터로 넘겨주면 안에서 구분문자를 꺼내서 사용하게 되었습니다.

* ##  pieces/PawnTest  class
이전에 사용하던 상수들을 enum으로 전부 변경하였습니다.

* ##  chess/Board  class
원래는 보드 하나당 pawns 하나를 가지고 있었지만 whitePawns, blackPawns 리스트 2개로 분리했습니다.  
그로인해 각각의 size나 find하는 메소드들이 생겨나 많은 메소드가 내부에 있습니다.  

1. #### initialize 메소드
    초기화 함수로 pawns 리스트들을 clear 해주고 8번 반복해서 pawn들을 넣어줍니다.

2. #### getPawnsResult 메소드
    pawns들에서 들어있는 pawn들을 알기 위해 사용합니다. 알고자 하는 색상을 매개변수로 받아서 분기점을 통해 getPawnsToString 을 호출해 String을 return 해줍니다.

3.  #### getPawnsToString 메소드
    pawns들 안에 들어있는 pawn들을 String형태로 반환해 줍니다.


* ##  chess/BoardTest class

한 개였던 pawns를 2개의 리스트(white, black)로 분리해서 많은 메소드들이 생성되었습니다.  

1. #### verifyPawnToPawns 메소드
    색상을 주어서 해당 색상의 pawn을 넣고 잘 들어갔는지 index를 통해 검사합니다. 내부적으로는 입력받은 enum을 통해 분기점을 만들어 verifyWhitePawnsIndex, verifyBlackPawnsIndex 메소드를 각각 호출해줍니다.  

2. #### verifyPaswsSize 메소드
    색상을 주어서 해당 pawns의 size가 증가하고 있는지 검증합니다. 매개변수로 받은 enum을 통해 분기점을 만들어 verifyWhitePaswsSize, verifyBlackPaswsSize 메소드를 각각 호출합니다.  

3. #### initialize 메소드와 print 메소드
     해당 메소드를 검증합니다.  



pawns 리스트가 색상대로 2개로 분리되어서 메소드가 2배로 증가하였습니다.   
긴 글 읽어주셔서 감사드립니다!

choigiseong committed:
honux 디렉토리 삭제

사용하지 않는 호눅스 디렉토리를 삭제했다.

choigiseong committed:
PawnEnum enum 작성

새로운 요구 사항인 색에 따른 구분 문자를 위해
Pawn class안에 있던 색상 상수를 이넘으로 교체했습니다.
색상 별로 구분 문자를 representation이라는 이름으로 넣었습니다.

choigiseong committed:
Pawn class에 enum을 도입

상수로 있던 색상을 enum으로 빼서
생성자를 새롭게 작성을 했습니다.
이넘을 사용했기에 생성할 때 pawn의 구분문자는 따로 주지 않아도
생성할 수 있게 되었습니다.

choigiseong committed:
Pawn class에서 getter 작성

각 필드는 getter를 통해 값을 얻을 수 있습니다.

choigiseong committed:
PawnTest class 에서 기본생성자 테스트

상수에서 enum으로 변경해서 상수 사용 부분을 enum으로 변경했다.
전에 확인해보니 기본 생성자 test코드가 없어서 이번에 새로 작성했다.

choigiseong committed:
PawnTest class에 colortest 추가

상수를 사용해 검사를 행해서 enum으로 변경했다.
추가로 구분 문자도 검사를 진행한다.

choigiseong committed:
Board class에서 pawn을 색상 별로 구분

원래는 하나의 리스트로 구분했지만, 색상별로 구분하기
위해 2개의 리스트로 분리했다.
기본 생성자는 리스트가 2개로 늘어남에 따라 생성자에서 각각 초기화
해주기 보다 필드에서 초기화를 해주는 게 좋을 거 같아 변경했다.

choigiseong committed:
addPawn을 색상별로 구분

리스트가 2개로 분리됨에 따라
각각 해당 Pawns에 넣어주는 메소드를 작성

choigiseong committed:
size와 findPawn을 색상별로 분리

pawns리스트가 2개의 리스트로 분리됨에 따라
각각 리스트의 크기와 찾는 index의 pawn을 리턴해주는
메소드를 작성했다.

choigiseong committed:
initialize메소드 작성

초기화를 위해 8번 반복해서 pawn을 넣어준다.

choigiseong committed:
getPawnsResult 메소드 작성

각각의 리스트에 들어있는 pawn들을 string으로 반환 해주기 위해 작성.
매개변수로 enum을 받아서 그것으로 판단해서 값을 리턴해준다.

choigiseong committed:
getPawnsToString 메소드 작성

Pawns를 받아서 내용물을 모두 붙여서 문자열로 반환
test에서 값들이 초기화가 잘 되었는지 검사하기 위해 작성되었다.

choigiseong committed:
BoardTest class create함수 변경

상수를 사용하다가 enum을 사용해서 각각 파라미터를 변경해주었다.
pawns 리스트가 whitePawns와 blackPawns로 나뉘었기에
size함수도 색상별로 구분했다.

choigiseong committed:
verifyPawnToPawns 함수 변경

리스트가 2개로 나뉘어서 분기점을 만들어
각각에 대응하는 함수를 호출하게 만들었다.

choigiseong committed:
verify...PawnsIndex 메소드 작성

색상별로 매개변수로 받은 인덱스에
원하는 pawn이 들어가 있는지 검사한다.

choigiseong committed:
verifyPaswsSize 메소드 작성

pawns가 색상별로 2개로 분리되어서 크기도 분기점을 통해
size를 검사한다.

choigiseong committed:
verify...PaswsSize

색상에 따른 size를 얻기위해 메소드를 작성했다.

choigiseong committed:
initialize 메소드 작성

색상별로 구분문자가 잘 들어 갔는지 검사한다.

choigiseong committed:
test에서 접근제어자 삭제

test에서는 접근지시자가 필요없어서 삭제했다.

choigiseong committed:
print 메소드 작성

빌더를 통해 체스판 모양을 만들어 string으로 반환한다.

choigiseong committed:
blank 메소드 작성

판에서 빈칸은 *을 찍어서 보여준다.

choigiseong committed:
print test 작성

게임 처음 시작 모습의 체스판을 출력한다.

choigiseong committed:
StartGame class 작성

사용자의 입력을 받아서 switch를 통해
정의되어 있는 행동을 취한다.

choigiseong committed:
commandSwitch 메소드 작성

input값을 switch로 판단해서 상황에 맞는 함수를 호출한다.

choigiseong committed:
initialize 메소드에 clear추가

보드판은 재활용할 수 있으니 pawn들은
재배치를 할 수 있어야한다고 생각했다.

tmdgusya commented on src/main/java/net/coco/pieces/PawnEnum.java:
```diff
@@ -0,0 +1,12 @@
+    WHITE_COLOR("p")
+    ,BLACK_COLOR("P");
+
+    PawnEnum(String representation) {
```
개인적으로 String 보다는 char 를 이용했으면 어땠을까 라는 생각이 듭니다.

choigiseong replied:
의견 감사합니다 로치! 문자 하나만 저장하기에 char로 다시 변경했습니다. 평소 char보다는 String을 자주 사용해 이런 부분은 생각지 못했네요.

tmdgusya commented on src/main/java/net/coco/pieces/PawnEnum.java:
```diff
@@ -0,0 +1,12 @@
+        this.representation = representation;
+    }
+
+    public String representation;
```
지금 상태에서 아래와 같은 코드를 외부에서 입력하면 어떻게 될까요

```java
PawnEnum. WHITE_COLOR.representation = 'c';
```

choigiseong replied:
외부에서 변경할 거라는 생각을 하지 못했네요. final을 붙여서 해결해 봤습니다.
이런 부분 하나하나가 후에 예기치 못한 장애를 만들 수 있겠다는 생각을 하는 좋은 기회였습니다.
감사합니다!

sonypark replied:
Enum의 필드 접근제어자를 public으로 두는게 좋은 선택인지 생각해보세요~
생성자와 필드의 위치도 확인 부탁드립니다.

choigiseong replied:
음.. 현재는 필드에 직접 접근하는 모습인데 접근제어자를 변경하고 getter 메소드를 추가하는 것으로 변경했습니다.
이러면 final도 굳이 붙일 필요가 없을 거 같아 final은 삭제 조치했습니다.

tmdgusya commented on src/main/java/net/coco/chess/Board.java:
```diff
@@ -70,4 +70,8 @@ public String print() {
         return boardPrint.toString();
     }
 
+    private String blank() {
```
해당 부분을 private final static blank = "********\n"; 로 했을때와 지금과 같이 method 를 이용했을때
어떤 차이가 있을까요? 어느 방법이 조금 더 지금상황에 좋은 방법일까요?

sonypark replied:
메서드로 빼기에는 역할이 너무 없는 것 같네요😅

choigiseong replied:
그렇네요. 상수를 사용하면 메소드 3줄에서 1줄로 코드랑도 줄어들어 좋은 거 같습니다.
매번 함수로 빼지 말고 역할을 다시 한 번 생각해 봐야겠네요

sonypark commented on src/main/java/net/coco/chess/Board.java:
```diff
@@ -1,24 +1,79 @@
+
+    public String print() {
+        StringBuilder boardPrint = new StringBuilder();
+        boardPrint.append(blank())
+                .append(getPawnsToString(whitePawns) + "\n")
+                .append(blank())
+                .append(blank())
+                .append(blank())
+                .append(blank())
+                .append(getPawnsToString(blackPawns) + "\n")
+                .append(blank());
```
체스말의 위치가 바뀐것 같네요. 체스판에서 검은말과 흰말의 위치를 확인해주세요~

choigiseong replied:
하하.. 변경 완료했습니다..

sonypark commented on src/main/java/net/coco/chess/Board.java:
```diff
@@ -1,24 +1,79 @@
+                .collect(Collectors.joining());
+    }
+
+    public String print() {
+        StringBuilder boardPrint = new StringBuilder();
```
체스판을 출력하는 로직은 Board에서 수행해야할 기능일까요?

choigiseong replied:
Board를 print 하는 기능은 새로운 class를 만들어 그 안에 할당했습니다.
이 때문에 print 메소드의 내부구조가 pawns의 내용물들을 문자열로 받아서 print해 주는 형식으로 바뀌게 되었습니다.

sonypark commented on src/main/java/net/coco/chess/Board.java:
```diff
@@ -1,24 +1,79 @@
-    public Board() {
-        this.pawns = new ArrayList<>();
+
+    public void addWhitePawn(Pawn pawn) {
+        whitePawns.add(pawn);
+    }
+
+    public void addBlackPawn(Pawn pawn) {
+        blackPawns.add(pawn);
+    }
```
의도하지 않은 color의 piece가 들어올 경우 검증 로직이 필요할 것 같네요

choigiseong replied:
여러 예외처리를 안 하고 있었네요. 앞으로 더 신경 쓰도록 하겠습니다. 감사합니다!

sonypark commented on src/main/java/net/coco/chess/Board.java:
```diff
@@ -1,24 +1,79 @@
+    public void initialize() {
+        whitePawns.clear();
+        blackPawns.clear();
+        for (int i = 0; i < 8; i++) {
```
매직넘버를 줄일 수 있는 방법은 무엇일까요?

choigiseong replied:
상수로 변경해 봤습니다! 상수로 변경을 해서 개발자의 의도를 쉽게 파악할 수 있고 만약에 경우 값들도 쉽게 변경할 수 있어 졌습니다.

sonypark commented on src/main/java/net/coco/chess/Board.java:
```diff
@@ -1,24 +1,79 @@
+    }
+
+    private String getPawnsToString(List<Pawn> Pawns) {
+        return Pawns.stream()
+                .map(Pawn::getRepresentation)
+                .collect(Collectors.joining());
```
👍

sonypark commented on src/main/java/net/coco/chess/StartGame.java:
```diff
@@ -0,0 +1,33 @@
+public class StartGame {
+
+    public static void main(String[] args) {
+        Scanner sc = new Scanner(System.in);
+        Board board = new Board();
```
Scanner 같은 객체는 생성한 뒤 close를 해줄 것을 권장합니다. 그 이유는 무엇일까요?

choigiseong replied:
close() 명령을 수행하였습니다. 명령을 선언해 주지 않으면 gc가 수거해 가지 않아 메모리 누수가 발생하는 것을 알았습니다. gc는 메모리 리소스만 관리하지 시스템 메모리 리소스는 개발자가 닫아야 하는군요. gc에 대해 좀 더 공부하겠습니다!

sonypark commented on src/main/java/net/coco/chess/StartGame.java:
```diff
@@ -0,0 +1,33 @@
+
+    private static boolean commandSwitch(Board board, String input) {
+        switch (input) {
+            case "start":
+                board.initialize();
+                System.out.println(board.print());
+                break;
+            case "end":
```
start, end와 같은 문자열은 상수를 활용해주세요~

choigiseong replied:
상수로 변경했습니다. 변하지 않는 값이기에 상수로 선언하는 게 맞는 거 같습니다!

sonypark commented on src/main/java/net/coco/pieces/PawnEnum.java:
```diff
@@ -0,0 +1,12 @@
+package net.coco.pieces;
+
+public enum PawnEnum {
+    WHITE_COLOR("p")
+    ,BLACK_COLOR("P");
```
PawnEnum을 만드신 이유는 무엇인가요?
만약 다른 Piece추가 된다면 Piece마다 Enum이 생기는걸까요?
PawnEnum의 역할이 무엇이고 왜 필요한지, 그렇지 않다면 어떻게 해야할지 고민해보시길 바랍니다.
네이밍에 자료형이 들어가는게 맞을지도 생각해보세요~

choigiseong replied:
맞는 말씀이십니다. PawnEnum으로 만들어 사용하면 확장성도 떨어지고 다른 Piece들이 늘어나면 다른 enum들이 늘어날 수 있습니다. 저 또한 이런 생각을 가지고 있었으나 이번 미션은 이렇게 해보고 다음 미션에서 만나는 요구사항에 따라 새롭게 고치는 것도 경험해 보고 싶었습니다. 
근데 지금 와서 생각해보니 representation에 대한 대처가 힘들 것 같다는 생각이 듭니다. representation에 관해서는 새로운 방법을 모색하거나 다른 분들의 코드를 참고해야 할 것 같습니다. 감사합니다.

sonypark commented on src/main/java/net/coco/chess/Board.java:
```diff
@@ -1,24 +1,79 @@
+
+    public int getWhitePawnSize() {
+        return whitePawns.size();
     }
 
-    public void addPawn(Pawn pawn){
-        pawns.add(pawn);
+    public int getBlackPawnSize() {
+        return blackPawns.size();
```
이 메서드는 필요한 기능인가요?

choigiseong replied:
저의 현재 구조는 white와 black, Pawns들의 size를 return 해주고 있습니다.
질문을 보고 다른 분들의 코드를 살펴보았습니다. 다른 분들은 두 리스트의 size를 합하여 return 하더군요.
코드의 목적은 pawn이 잘 들어가고 있는지 size를 검사 하는 거기 때문에 굳이 색상별로 나눌 필요가 없다는 것을 알게 되었습니다.

그래서 저 또한 get Size() 메소드로 합쳐 진행해 보겠습니다.

sonypark commented on src/test/java/net/coco/pieces/pawnTest.java:
```diff
@@ -1,21 +1,37 @@
+        verifyPawn(PawnEnum.BLACK_COLOR);
+    }
 
+    void verifyPawn(PawnEnum pawnEnum) {
+        Pawn pawn = new Pawn(pawnEnum);
+        Assertions.assertAll(
+                () -> assertThat(pawnEnum).isEqualTo(pawn.getColor()),
+                () -> assertThat(pawnEnum.representation).isEqualTo(pawn.getRepresentation())
+        );
     }
```
👍

sonypark commented on src/test/java/net/coco/chess/BoardTest.java:
```diff
@@ -14,32 +15,72 @@
+
+    }
+
+    private void verifyPaswsSize(PawnEnum pawnEnum, int actualSize) {
+        if (pawnEnum == PawnEnum.WHITE_COLOR)
+            verifyWhitePaswsSize(actualSize);
+        else
+            verifyBlackPaswsSize(actualSize);
+    }
```
Pawn 스펠링~🔤

choigiseong replied:
수정했습니다!

sonypark commented on src/main/java/net/coco/chess/Board.java:
```diff
@@ -1,24 +1,79 @@
+        whitePawns.clear();
+        blackPawns.clear();
+        for (int i = 0; i < 8; i++) {
+            whitePawns.add(new Pawn(PawnEnum.WHITE_COLOR));
+            blackPawns.add(new Pawn(PawnEnum.BLACK_COLOR));
+        }
```
이 부분은 addWhitePawn, addBlackPawn과 중복되네요. 만들어 놓은 메서드를 활용해보세요

choigiseong replied:
수정했습니다! 이미 있는 함수를 사용하니 코드의 목적이 더 잘 보이는 것 같습니다!

sonypark reviewed (changes requested):
안녕하세요 coco, 리뷰어 소니 입니다.😊
요구사항에 맞게 기능이 잘 동작하네요.
다만 몇가지 아쉬운 점이 있어 코멘트 남겼으니 확인 후 수정 부탁드릴게요!

choigiseong committed:
Pawn의 representation 자료형 변경

하나의 문자이기 때문에 String보다는 char를 쓰는게
좋을거 같다는 피드백을 받아 변경했습니다.
구분문자이기에 후에도 2문자 이상은 들어오지 않으꺼 같아
변경했습니다.

choigiseong committed:
blank를 메소드에서 상수로 변경

메소드로 사용을 했을 때는 3줄이었는데
상수로 사용을 하니 1줄로 변경이 되었습니다.

choigiseong committed:
Board class에서 기능 분할

Board class안에 있던 print를 PrintChess class로 옮겼습니다.
체스 판을 print해주는 역할은 Board에 있기보단 print만 담당하는
class를 새로 생성해 줬습니다.

choigiseong committed:
add...pawn 메소드에 검증 로직 추가

이전에는 색상이 맞지 않은 pawn이 들어와도
add되었지만 지금은 검증을 통해 알맞게 넣습니다.

choigiseong committed:
매직넘버를 상수로 변환

상수로 변경해서 쉽게 의도를 파악할 수 있고
만약에 경우 값들도 쉽게 변경이 가능해 졌다.

choigiseong committed:
Scanner close 수행

close() 메소드를 사용하지 않았는데, 이번에는 close()를 사용해
메모리 리소스의 누수를 막을 수 있게 됬다.

choigiseong committed:
게임 명령어를 상수로 변경

start와 end를 상수로 변경해 의도를 쉽게 파악하게 했습니다.

choigiseong committed:
PawnEnum에서 상수 이름 변경

enum뒤에 _color 를 붙여 사용했었는데
삭제 조치를 했다. 색을 표시 안해줘도 pawn이기 때문에
색상인 것을 알 수 있다.

choigiseong committed:
get...PawnSize메소드를 병합

각기 다른 색상의 크기를 return해주는 구조는 좋지 않다고 생각했다.
목적은 pawn이 잘 들어가고 있는지 확인하는게 목적이기에
둘의 size를 합쳐서 return해주는 쪽으로 변경을 했다.

choigiseong committed:
오타 수정 및 메소드 수정

Pawn 오타를 수정했습니다. 또 size()메소드를 합침에 따라 불필요해진 메소드를
삭제하였습니다.
initialize()에서 만들어 놓은 메소드를 사용하게 했습니다.

choigiseong committed:
add...Pawn 메소드 병합

색상별로 메소드가 존재하였지만 하나의
메소드에 몰아넣었습니다.
이 때문에 코드량이 줄었고 개발자는
하나의 add 메소드만 사용하면 되게 되었습니다.

sonypark commented on src/main/java/net/coco/chess/StartGame.java:
```diff
@@ -0,0 +1,45 @@
+            case START:
+                board.initialize();
+                String blackPawns = board.getPawnsResult(PawnEnum.BLACK);
+                String whitePawns = board.getPawnsResult(PawnEnum.WHITE);
+                System.out.println(PrintChess.printBoard(blackPawns, whitePawns));
```
PrintChess 클래스의 `printBoard()` 메서드를 보면 출력을 담당할 것 같은데 출력은 `commandSwitch()`에서 해주고 있네요.

choigiseong replied:
printBoard()의 형태를 결과물만 return하는 모습으로 설계를 했었는데, 
printBoard() 메소드에서 출력까지 하는 것으로 변경했습니다!

sonypark commented on src/main/java/net/coco/chess/StartGame.java:
```diff
@@ -0,0 +1,45 @@
+            case END:
+                System.out.println("게임을 종료");
+                return true;
+            default:
+                throw new IllegalArgumentException("알 수 없는 명령어");
+        }
```
명령어를 잘못 입력했을 때 프로그램을 종료시키는 것이 자연스러운 흐름일까요?

choigiseong replied:
사용자로서는 다시 입력받는 게 좋을 거 같아 변경했습니다!

sonypark reviewed (changes requested):
수정이 필요한 곳에 코멘트 남겼으니 반영 부탁드릴게요~

choigiseong committed:
PawnEnum의 representation 접근제어자 변경

private로 변경해서 외부에서 접근하려면
getter를 통해 접근하도록 바꿨습니다.
또 final은 접근제어자를 변경함에 따라 필요가
없을 거 같아 삭제했습니다.

choigiseong committed:
PrintChess에서 출력하도록 변경

String을 return하는 형태가 아닌 출력까지 여기서
모두 하게 변경

choigiseong committed:
commandSwitch 메소드 예외처리 변경

알 수 없는 명령어 입력 시 게임을 종료시켰는데,
사용자로서는 다시 입력받는 게 좋을 거 같아 변경했습니다

sonypark commented on src/main/java/net/coco/chess/StartGame.java:
```diff
@@ -31,7 +31,7 @@ private static boolean commandSwitch(Board board, String input) {
                 String blackPawns = board.getPawnsResult(PawnEnum.BLACK);
                 String whitePawns = board.getPawnsResult(PawnEnum.WHITE);
-                System.out.println(PrintChess.printBoard(blackPawns, whitePawns));
+                PrintChess.printBoard(blackPawns, whitePawns);
```
`printBoard()`라는 메서드 이름에 맞게 인자로 board를 넘기는건 어떨까요?

choigiseong replied:
printBoard() 하기 전에 파라미터를 준비하는 과정이 있어서 코드량이 길어졌는데, 그냥 보드를 넘기면 됐었네요.
왜 이런 생각을 못했을까요. 감사합니다!

sonypark commented on src/main/java/net/coco/pieces/PawnEnum.java:
```diff
@@ -1,12 +1,15 @@
-    WHITE('p')
-    , BLACK('P');
+    WHITE('p'), BLACK('P');
+
```
Enum의 필드는 한 줄에 하나씩 구분하여 명시하도록 해주세요~
```java
PawnEnum {
    WHITE('p'), 
    BLACK('P');
```

sonypark commented on src/main/java/net/coco/pieces/Pawn.java:
```diff
@@ -1,16 +1,24 @@
+    private char representation;
 
-    private String color;
+    public Pawn() {
+        this.color = PawnEnum.WHITE;
+        this.representation = PawnEnum.WHITE.getRepresentation();
+    }
 
-    public Pawn(String color) {
-        this.color = color;
+    public Pawn(PawnEnum pawnEnum) {
+        this.color = pawnEnum;
+        this.representation = pawnEnum.getRepresentation();
     }
 
-    public String getColor() {
+    public PawnEnum getColor() {
         return color;
     }
```
color의 타입이 PawnEnum인게 맞을까요? color와 representation은 하나의 Enum으로 묶는게 적절한지도 고민이 필요해 보입니다. Enum을 사용하기 어렵다면 일단 자바 상수를 이용해서 구현하는 걸 추천드립니다. Enum을 잘 모르는 상태에서 오용하게 되면 오히려 코드의 복잡도를 높일 수 있습니다. 먼저 자바 상수로 구현한 후 추후 미션에서 Enum으로 변경하는 것도 좋은 방법입니다.

choigiseong replied:
조언 감사드립니다. 저의 의도는 현재는 pawn만 있으니 미래를 생각하지 않고 pawn에만 딱 들어맞게 설계를 했습니다. PawnEnum을 생각해보면 확장성은 없고 pawn의 특징들만 가지고 있습니다. color라든지 'p'와 'P'로 정의된 representation 라던지.. 
둘을 묶은 이유는 파라미터를 간단히 넘겨주고 싶은 마음도 있었습니다. 
하지만 이런 생각 때문에 코드의 복잡도 생각은 하지 못했네요. Pawn의 생성자는 겉으로는 enum만 파라미터로 넘겨주는데, 내부에서는 enum 내부에 있는 representation까지 사용하니 외부에서 봤을 때 정확한 동작을 예측하기 힘들 거 같다는 생각도 듭니다.
아직 다음 미션을 보지는 않았지만 미션4를 확인하고 다른 enum으로  representation 을 빼든지 아니면 기물마다 구분문자를 상수로 가지고 있게 하던지 설계를 다시 진행하겠습니다. 감사합니다.

sonypark commented on src/main/java/net/coco/pieces/PawnEnum.java:
```diff
@@ -0,0 +1,15 @@
+package net.coco.pieces;
+
+public enum PawnEnum {
+    WHITE('p'), BLACK('P');
```
```suggestion
public enum PawnEnum {
    WHITE('p'), 
    BLACK('P');
```
Enum의 필드는 한 줄에 하나씩 입력해주세요~

choigiseong replied:
수정 완료했습니다!

sonypark reviewed (changes requested):
수정해주신 부분 확인했습니다!
코멘트 남긴 부분 확인해주세요~
특히 Enum 사용 관련한 부분에 대해 수정이 필요해보입니다.🤔

choigiseong committed:
printBoard 메소드 매개변수 변경

원래는 문자열 형태로 각각의 pawn들을 받아서 사용을 했는데,
board자체를 받고 안에서 활용하는 쪽으로 변경을 했습니다.
이러면 printBoard() 메소드를 사용전에 파라미터를 준비하는 과정을
줄일 수 있어 좋습니다.

sonypark reviewed (approved):
확인했습니다.
다음 단계 진행해주세요~💪

honux77 commented:
수고하셨어요!
